symforce/slam/imu_preintegration/imu_factor.h

```diff
@@ -0,0 +1,82 @@
+/* ----------------------------------------------------------------------------
+ * SymForce - Copyright 2022, Skydio, Inc.
+ * This source code is under the Apache 2.0 license found in the LICENSE file.
+ * ---------------------------------------------------------------------------- */
+
+#pragma once
+
+#include <Eigen/Dense>
+
+#include <sym/pose3.h>
+
+#include "./preintegrated_imu_measurements.h"
+
+namespace sym {
+
+/**
+ * A Functor storing on-manifold preintegrated IMU measurements between two times and
+ * which computes the local-coordinate difference between the final state (pose and
+ * velocity) and the state you would expect given the initial state and
+ * the IMU measurements.
+ *
+ * Based heavily on the on manifold ImuFactor found in GTSAM and on the paper:
+ *
+ * Christian Forster, Luca Carlone, Frank Dellaert, and Davide Scaramuzza,
+ * “IMU Preintegration on Manifold for Efficient Visual-Inertial Maximum-a-Posteriori Estimation”,
+ * Robotics: Science and Systems (RSS), 2015.
+ */
+template <typename Scalar>
+class ImuFactor {
+ private:
+  PIMNoCovariance<Scalar> pim_;
+  Eigen::Matrix<Scalar, 9, 9> sqrt_info_;
+
+ public:
+  /**
+   * Construct an ImuFactor connecting two states from the (preintegrated) imu measurements
+   * between them.
+   */
+  ImuFactor(const PreintegratedImuMeasurements<Scalar>& pim);
+
+  /**
+   * Calculate a between factor on the product manifold of the pose and velocity where the prior
+   * is calculated from the preintegrated IMU measurements.
+   *
+   * Time step i is the time of the first IMU measurement of the interval.
+   * Time step j is the time after the last IMU measurement of the interval.
+   *
+   * Args:
+   *     pose_i: Pose at time step i
+   *     vel_i: Velocity at time step i
+   *     pose_j: Orientation at time step j
+   *     vel_j: Velocity at time step j
+   *     accel_bias_i: The bias of the accelerometer measurements between timesteps i and j
+   *     gyro_bias_i: The bias of the gyroscope measurements between timesteps i and j
+   *     gravity: Acceleration due to gravity
```

Comment:
The gravity here is in whatever fixed frame the poses are in. However, what I meant was we should mention that we assume gravity in the world frame is [0, 0, -g].

Reply:
Makes sense, I think we should probably mention both of those things here

Reply:
Perhaps this is because I'm not used to the convention here, but saying that the gravity is assumed to be [0, 0, -g] or [0, 0, g] is confusing to me as it makes the unnecessary assumption that the world frame is (anti-)gravity aligned. Moreover, it kind of just seems unrelated to the core question to me? Specifically, I was thinking the important question is what assumption is being made about the readings of the IMU accelerometer (what would it read while stationary and/or on a ballistic trajectory).

So, this is what I added to the doc-string:

```
 * Assumes that an IMU falling only under the influence of gravity will measure 0 acceleration;
 * the gravity argument to the factor is expected to be the true acceleration of such an IMU in
 * the world frame.
```
For example, if the IMU on a ballistic trajectory did measure acceleration relative to Earth, i.e., 0 if stationary, then one should pass gravity as 0 to the factor.

Reply:
That's all fair, and it's fine to call out all of those things - but as someone looking to use this, 99.999% of people are going to be using a frame that is gravity-aligned and fixed relative to the earth's surface, and are just asking the question "should this be `[0, 0, g]` or `[0, 0, -g]`".  I strongly think we should include what you should pass for this parameter in that case, so 99.999% of people don't have to read this entire thing and think through what it means to deduce what they need to pass here.

The core question is indeed "is this vector the acceleration measured by the IMU at rest, or the acceleration you need to add to that to get 0".

Reply:
Okay. I tried to make it easier for a user who wants to figure out what argument to pass in without having to think that hard about it:
```
 * The gravity argument should be [0, 0, -g] (where g is 9.8, assuming your world frame is
 * gravity aligned so that the -z direction points towards the Earth) unless your IMU reads 0
 * acceleration while stationary, in which case it should be [0, 0, 0].
 *
 * More generally, the gravity argument should yield the true acceleration when added to the
 * accelerometer measurement (after the measurement has been converted to the world frame).
```

Reply:
Would it be easier to just write out the accel measurement equation?
```
acc_body = body_R_fixed * (acc_fixed - g_fixed)
```
I use `fixed` because this frame doesn't have to be the (gravity-aligned) `world` frame. It can be any fixed frame as long as you know the gravity vector in that frame.

Reply:
> acc_body = body_R_fixed * (acc_fixed - g_fixed)

I agree this is both more concise and easier to read imo, I'm fine either way